src/NuGet.Core/NuGet.Configuration/PackageSource/PackageSourceProvider.cs

```diff
@@ -298,178 +240,472 @@ private PackageSourceCredential ReadCredentialFromEnvironment(string sourceName)
+
+                    if (sourceMatching || feedNameMatching)
                     {
-                        packageSourcesToBeRemoved.Add(loadedPackageSources[i]);
+                        return packageSource;
                     }
-                    else
+                }
+            }
+
+            return null;
+        }
+
+        public PackageSource GetPackageSourceWithName(string name)
+        {
+            if (string.IsNullOrEmpty(name))
+            {
+                throw new ArgumentNullException(nameof(name));
+            }
+
+            return GetPackageSource(name, LoadPackageSourceLookupByName());
+        }
+
+        public PackageSource GetPackageSourceWithSource(string source)
+        {
+            if (string.IsNullOrEmpty(source))
+            {
+                throw new ArgumentNullException(nameof(source));
+            }
+
+            return GetPackageSource(source, LoadPackageSourceLookupBySource());
+        }
+
+        public void RemovePackageSource(string name)
+        {
+            if (string.IsNullOrEmpty(name))
+            {
+                throw new ArgumentNullException(nameof(name));
+            }
+
+            var isDirty = false;
+            RemovePackageSource(name, shouldSkipSave: false, isDirty: ref isDirty);
+        }
+
+        private void RemovePackageSource(string name, bool shouldSkipSave, ref bool isDirty)
+        {
+            // get list of sources
+            var packageSourcesSection = Settings.GetSection(ConfigurationConstants.PackageSources);
+            var sourcesSettings = packageSourcesSection?.Items.Select(s => s as SourceItem)
+                .Where(s => s != null);
+
+            // get list of credentials for sources
+            var sourceCredentialsSection = Settings.GetSection(ConfigurationConstants.CredentialsSectionName);
+            var sourceCredentialsSettings = sourceCredentialsSection?.Items.Select(s => s as CredentialsItem)
+                .Where(s => s != null);
+
+            var sourcesToRemove = sourcesSettings?.Where(s => string.Equals(s.Key, name, StringComparison.OrdinalIgnoreCase));
+            var credentialsToRemove = sourceCredentialsSettings?.Where(s => string.Equals(s.Name, name, StringComparison.OrdinalIgnoreCase));
+
+            if (sourcesToRemove != null)
+            {
+                foreach (var source in sourcesToRemove)
+                {
+                    try
                     {
-                        loadedPackageSources[i] = targetPackageSource.Clone();
-                        // make sure we preserve the IsEnabled property when migrating package sources
-                        loadedPackageSources[i].IsEnabled = ps.IsEnabled;
+                        Settings.Remove(ConfigurationConstants.PackageSources, source);
+                        isDirty = true;
                     }
-                    hasChanges = true;
+                    catch { }
```

Comment:
This is to preserve previous behavior... This will fail when the user is trying to remove a machine-wide source. The previous behavior was to swallow any error and just ignore the user's request. Since this PR is focused on config changes and not sources changes, I didn't make any functional changes to this class.